core/kotlinx-coroutines-core/src/EventLoop.kt

```diff
@@ -318,27 +311,31 @@ internal abstract class EventLoopBase: CoroutineDispatcher(), Delay, EventLoop {
     }
 }
 
-internal abstract class ThreadEventLoop(
-    private val thread: Thread
-) : EventLoopBase() {
-    override fun isCorrectThread(): Boolean = Thread.currentThread() === thread
+internal class BlockingEventLoop(
+    override val thread: Thread
+) : EventLoopImplBase()
 
-    override fun unpark() {
-        if (Thread.currentThread() !== thread)
-            timeSource.unpark(thread)
-    }
+internal actual fun createEventLoop(): EventLoop = BlockingEventLoop(Thread.currentThread())
 
-    fun shutdown() {
-        closeQueue()
-        assert(isCorrectThread())
-        // complete processing of all queued tasks
-        while (processNextEvent() <= 0) { /* spin */ }
-        // reschedule the rest of delayed tasks
-        rescheduleAllDelayed()
-    }
-}
-
-internal class BlockingEventLoop(thread: Thread) : ThreadEventLoop(thread) {
-    @Volatile
-    public override var isCompleted: Boolean = false
-}
+/**
+ * Processes next event in the current thread's event loop.
+ *
+ * The result of this function is to be interpreted like this:
+ * * `<= 0` -- there are potentially more events for immediate processing;
+ * * `> 0` -- a number of nanoseconds to wait for the next scheduled event;
+ * * [Long.MAX_VALUE] -- no more events, or was invoked from the wrong thread.
```

Comment:
This one is hard to fix. The problem is that you can do `runBlocking(context) { ... }` and if context has `EventLoop` inside, then `runBlocking` works in that event loop (if and only if the thread is right). We need to discuss whether this use-case is needed in the first-place, and then this contract can be maybe changed.

Reply:
I've found that original behavior of `runBlocking(context)` when context has `EventLoop` was poorly thought-though and would have blocked thread's event loop. Fixed it and added test. Now `processNextEvent` would only be called from the correct thread and since it is an internal API, I've removed the check (for performance reasons) and rephrased the docs.